urlcaef331f4e_dreamable_cra_tgz-pJ8-coho_lp_project_LPProjectJ8/src/coho/geom/twodim/ops/IntersectPoint.java

```diff
@@ -0,0 +1,382 @@
+		}
+		//If it is the endpoint of s1 and s2, but it has more than 2 intersection segements, it's intersection point
+		//System.out.println(s1.getSegmentsNumberOfPoint(this));
+		if(s1.getSegmentsNumberOfPoint(this)>2)
+			return true;
+		//If it is the endpoint of s1 and s2, and has only to intersectio segments, it must be a vertex on only one polgon.
+		else
+			return  false;
+	}
+	// forward endpoint, can't go further.
+	public boolean isFwdEndPoint(Segment seg){
+		return dynamicCompare(seg.p(1))==0;
+	}
+	public boolean isEndPoint(Segment seg){
+		return ((dynamicCompare(seg.p(0))==0) || (dynamicCompare(seg.p(1))==0));
+	}
+	
+	
+	/******************************
+	 * Compare functions. In the intersection step, we don't care the order of points, but at the tranverse
+	 * step, the point should be ordered by their coordinate to find the union/intersect. The points with 
+	 * same coordinates are "virtually combined".  
+	 ******************************/
+	/*
+	 * Compare the intersection point by its x-y coordinate. If it's not comparable, re-evaluation
+	 * is performed to get the exact value. 
+	 */
+	public int dynamicCompare(Point that){
+		if(that instanceof IntersectPoint){
+			return dynamicCompare((IntersectPoint)that);
+		}
+		try{
+			return point().compareTo(that);
+		}catch(ArithmeticException e){
+			//CONSIDER: avoid loop, Consider divide-by-zero error for apr representation.
+			//DONE: But there is no divide computation for compareTo function. Other exception?
+//			if(exact)
+//				throw e;
+			eval();
+			return aprPoint.compareTo(that);
+		}
+	}
+	/*
+	 * Compare the intersection point by its x-y coordinate. If it's not comparable, re-evaluation
+	 * is performed to get the exact value. 
+	 */
+	public int dynamicCompare(IntersectPoint that){
+		int result ;
+		try{
+			//System.out.println("compare "+this+ "and" + that.point()+" type "+this.type()+that.type());
+			result = point().compareTo(that.point());//CONSIDER: Can we merge here? Highly possible, do it later.
+			//System.out.println("1 "+result+this.type()+that.type());
+		}catch(ArithmeticException e){
+			//CONSIDER: avoid loop, Consider divide-by-zero error for apr representation.
+			//DONE: But there is no divide computation for compareTo function. Other exception?
+//			if(exact && that.exact)
+//				throw e;
+			eval();
+			that.eval();
+			result = aprPoint.compareTo(that.aprPoint);
+		}
+		//System.out.println("result2 "+result);
+		return result;
+	}
+	/*
+	 * Recompute the intersection point using APR. This should be called rarely. 
+	 */
+	public void eval(){
+		if(exact)
+			return;
+		//NOTE: the intersection is impossible to be a segment. 
+		//This is because if intersection is segment, APR must be used before.
+		//aprPoint = (Point)s1.intersect(s2,true);//compute using APR.
+		try{
+			aprPoint = (Point)s1.specifyType(CohoAPR.type).intersectSegment(s2.specifyType(CohoAPR.type));//using apr
+			exact = true;
+		}catch(ClassCastException e){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)